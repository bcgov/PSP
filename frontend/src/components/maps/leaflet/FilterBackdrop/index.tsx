import * as React from 'react';
import Spinner from 'react-bootstrap/Spinner';
import styled from 'styled-components';

export interface FilterBackdropProps {
  show?: boolean;
}

const Backdrop = styled.div`
  width: 100%;
  height: 100%;
  position: absolute;
  z-index: 999;
  left: 0;
  background-color: rgba(0, 0, 0, 0.4);
  display: flex;
  align-items: center;
  align-content: center;
  justify-items: center;
  justify-content: center;
`;

const FilterBackdrop: React.FC<FilterBackdropProps> = ({ show }) => {
  return show ? (
    <Backdrop>
      <Spinner animation="border" variant="warning" data-testid="filter-backdrop-loading" />
    </Backdrop>
  ) : null;
};

export default FilterBackdrop;
