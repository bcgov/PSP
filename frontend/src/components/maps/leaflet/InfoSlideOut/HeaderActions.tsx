import variables from '_variables.module.scss';
import { PropertyTypes } from 'constants/propertyTypes';
import { IProperty } from 'interfaces';
import * as React from 'react';
import Row from 'react-bootstrap/Row';
import { Link, useLocation } from 'react-router-dom';
import styled from 'styled-components';

const LinkMenu = styled(Row)`
  background-color: ${variables.filterBackgroundColor};
  height: 3.5rem;
  width: 32.2rem;
  margin: 0rem 0rem 0.5rem -1rem;
  font-size: 1.4rem;
  padding: 1rem;
  a {
    padding: 0rem 1rem;
    color: ${variables.slideOutBlue};
  }
`;

interface IHeaderActions {
  /** The selected property */
  propertyInfo: IProperty | null;
  /** the selected property type */
  propertyTypeId: PropertyTypes | null;
  zoomToView: () => void;
  /** additional action to be taken when a link in the menu is clicked */
  onLinkClick?: () => void;
  /** whether the user has the correct organization/permissions to view all the details */
  canViewDetails: boolean;
  /** whether the user has the correct organization/permissions to edit property details */
  canEditDetails: boolean;
}

/**
 * Actions that can be done on the property info slide out
 * @param propertyInfo the selected property information
 * @param propertyTypeId the selected property type
 * @param onLinkClick additional action on menu item click
 * @param canViewDetails user can view all property details
 * @param canEditDetails user can edit property details
 */
const HeaderActions: React.FC<IHeaderActions> = ({
  propertyInfo,
  propertyTypeId,
  onLinkClick,
  zoomToView,
  canViewDetails,
  canEditDetails,
}) => {
  const location = useLocation();

  return (
    <LinkMenu>
      Actions:
      <Link to={{ ...location }} onClick={zoomToView}>
        Zoom map
      </Link>
    </LinkMenu>
  );
};

export default HeaderActions;
