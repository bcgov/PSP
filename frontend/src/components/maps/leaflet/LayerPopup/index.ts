export * from './constants';
export * from './hooks/useLayerQuery';
export * from './LayerPopup';
export * from './LayerPopupContent';
