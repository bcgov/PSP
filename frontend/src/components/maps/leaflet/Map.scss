@import '../../../variables';

.leaflet-container {
  /* Subtract the header, nav-bar and footer sizes */
  height: calc(100vh - #{$header-height} - #{$mapfilter-height} - #{$footer-height});
  // margin-top: 1.0rem;
  // padding-top: 2.0rem;
  margin: auto;

  .leaflet-popup-content-wrapper {
    border-radius: 0;
  }
  .leaflet-popup-content {
    margin: 0rem;
    .list-group {
      .list-group-item {
        border: none;
        padding: 0.8rem 1.4rem;
      }
    }
  }

  .leaflet-control-container .leaflet-top {
    z-index: 500;
  }
}

.leaflet-control-layers {
  &.leaflet-control-layers-expanded {
    border-radius: 0rem;
  }
  .leaflet-control-layers-overlays {
    display: flex;
    flex-direction: column;
    padding: 1rem;
    width: 15rem;
    label {
      display: flex;
    }
  }
}
