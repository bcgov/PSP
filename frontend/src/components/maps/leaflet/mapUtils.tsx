import { DraftCircleNumber } from 'features/properties/selector/components/DraftCircleNumber';
import { IProperty } from 'interfaces';
import L, { DivIcon, GeoJSON, LatLngExpression, Layer, Map, Marker } from 'leaflet';
import React from 'react';
import ReactDOMServer from 'react-dom/server';
import Supercluster from 'supercluster';

import { ICluster, PointFeature } from '../types';

// parcel icon (green)
export const parcelIcon = L.icon({
  iconUrl: require('assets/images/pins/land-reg.png').default ?? 'assets/images/pins/land-reg.png',
  shadowUrl: require('assets/images/pins/marker-shadow.png').default ?? 'marker-shadow.png',
  iconSize: [25, 41],
  iconAnchor: [12, 41],
  popupAnchor: [1, -34],
  shadowSize: [41, 41],
});

// parcel icon (green) highlighted
export const parcelIconSelect = L.icon({
  iconUrl:
    require('assets/images/pins/land-reg-highlight.png').default ??
    'assets/images/pins/land-reg-highlight.png',
  shadowUrl: require('assets/images/pins/marker-shadow.png').default ?? 'marker-shadow.png',
  iconSize: [25, 41],
  iconAnchor: [12, 41],
  popupAnchor: [1, -34],
  shadowSize: [41, 41],
});

/**
 * Creates map points (in GeoJSON format) for further clustering by `supercluster`
 * @param properties
 */
export const createPoints = (properties: IProperty[], type: string = 'Point') =>
  properties.map(x => {
    return {
      type: 'Feature',
      properties: {
        ...x,
        cluster: false,
        PROPERTY_ID: x.id,
      },
      geometry: {
        type: type,
        coordinates: [x.longitude, x.latitude],
      },
    } as PointFeature;
  });

/**
 * This function defines how GeoJSON points spawn Leaflet layers on the map.
 * It is called internally by the `GeoJSON` leaflet component.
 * @param feature
 * @param latlng
 */
export const pointToLayer = (feature: ICluster, latlng: LatLngExpression): Layer => {
  const { cluster: isCluster } = feature?.properties as Supercluster.ClusterProperties;
  if (!!isCluster) {
    return createClusterMarker(feature, latlng);
  }
  // we have a single point to render
  return createSingleMarker(feature, latlng);
};

/**
 * Get an icon type for the specified cluster property details (type, draft, erp, spp etc)
 */
export const getMarkerIcon = (feature: ICluster, selected?: boolean) => {
  if (selected) {
    return parcelIconSelect;
  }
  return parcelIcon;
};

// parcel icon (green) highlighted
export const getDraftIcon = (text: string) => {
  return L.divIcon({
    iconSize: [29, 45],
    iconAnchor: [15, 45],
    popupAnchor: [1, -34],
    shadowSize: [41, 41],
    html: ReactDOMServer.renderToStaticMarkup(<DraftCircleNumber text={text} />),
  });
};

/**
 * Creates a map pin for a single point; e.g. a parcel
 * @param feature the geojson object
 * @param latlng the point position
 */
export const createSingleMarker = (feature: ICluster, latlng: LatLngExpression): Layer => {
  const icon = getMarkerIcon(feature);
  return new Marker(latlng, { icon });
};

// Internal cache of cluster icons to avoid re-creating the same icon over and over again.
const iconsCache: Record<number, DivIcon> = {};

/**
 * Creates a marker for clusters on the map
 * @param feature the cluster geojson object
 * @param latlng the cluster position
 */
export const createClusterMarker = (feature: ICluster, latlng: LatLngExpression): Layer => {
  const {
    cluster: isCluster,
    point_count: count,
    point_count_abbreviated: displayValue,
  } = feature?.properties as Supercluster.ClusterProperties;

  if (!isCluster) {
    return (null as unknown) as Layer;
  }

  const size = count < 100 ? 'small' : count < 1000 ? 'medium' : 'large';
  let icon: DivIcon;

  if (!iconsCache[count]) {
    iconsCache[count] = new DivIcon({
      html: `<div><span>${displayValue}</span></div>`,
      className: `marker-cluster marker-cluster-${size}`,
      iconSize: [40, 40],
    });
  }

  icon = iconsCache[count];
  return new Marker(latlng, { icon });
};

/** Zooms to a cluster */
export const zoomToCluster = (cluster: ICluster, expansionZoom: number, map: Map) => {
  const latlng = GeoJSON.coordsToLatLng(cluster?.geometry?.coordinates as [number, number]);
  map?.setView(latlng, expansionZoom, { animate: true });
};

// we need to namespace the keys as IDs are not enough here.
// the same ID could be found on both the parcel collection and
export const generateKey = (p: IProperty) => `parcel-${p.id}`;

/** Creates a IProperty object from a GeoJSON point */
export const asProperty = (point: PointFeature): IProperty => {
  const { id, name } = point?.properties;
  const latlng = GeoJSON.coordsToLatLng(point?.geometry?.coordinates as [number, number]);
  return {
    ...point.properties,
    id,
    latitude: latlng.lat,
    longitude: latlng.lng,
    name,
  } as IProperty;
};

/**
 * Convert any object to a cql filter string, assuming the object's keys should be used as CQL filter properties.
 * AND all object keys together within the generated cql filter string.
 *
 * The resulting filter is URL-encoded
 * @param object an object to convert to a cql filter string.
 */
export const toCqlFilter = (object: any, pidOverride?: boolean) => {
  const cqlValue: string = toCqlFilterValue(object, pidOverride);
  return cqlValue.length ? `cql_filter=${encodeURIComponent(cqlValue)}` : '';
};

/**
 * Convert any object to a cql filter string value, assuming the object's keys should be used as CQL filter properties.
 * NOTE: The resulting string value is left as-is. It will not be URL-encoded
 * @param object an object to convert to a cql filter string.
 */
export const toCqlFilterValue = (object: any, pidOverride?: boolean) => {
  const cql: string[] = [];
  Object.keys(object).forEach((key: string) => {
    if (object[key]) {
      if (key === 'PID' || (key === 'PID_PADDED' && object[key] && !pidOverride)) {
        cql.push(`PIN ilike '%${object[key]}%' OR PID ilike '%${object[key]}%'`);
      } else {
        cql.push(`${key} ilike '%${object[key]}%'`);
      }
    }
  });
  return cql.length ? cql.join(' AND ') : '';
};
