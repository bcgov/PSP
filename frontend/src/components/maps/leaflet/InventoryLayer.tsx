import { IGeoSearchParams } from 'constants/API';
import { PropertyTypes } from 'constants/propertyTypes';
import { BBox, Feature } from 'geojson';
import { useApi } from 'hooks/useApi';
import useDeepCompareEffect from 'hooks/useDeepCompareEffect';
import { geoJSON, LatLngBounds } from 'leaflet';
import { uniqBy } from 'lodash';
import React, { useEffect, useMemo, useState } from 'react';
import { useMap } from 'react-leaflet';
import { toast } from 'react-toastify';
import { useAppSelector } from 'store/hooks';
import { IPropertyDetail } from 'store/slices/properties';
import { tilesInBbox } from 'tiles-in-bbox';

import { useMapRefreshEvent } from '../hooks/useMapRefreshEvent';
import { useFilterContext } from '../providers/FIlterProvider';
import { PointFeature } from '../types';
import PointClusterer from './PointClusterer';

export type InventoryLayerProps = {
  /** Latitude and Longitude boundary of the layer. */
  bounds: LatLngBounds;
  /** Zoom level of the map. */
  zoom: number;
  /** Minimum zoom level allowed. */
  minZoom?: number;
  /** Maximum zoom level allowed. */
  maxZoom?: number;
  /** Search filter to apply to properties. */
  filter?: IGeoSearchParams;
  /** Callback function to display/hide backdrop*/
  onRequestData: (showBackdrop: boolean) => void;
  /** What to do when the marker is clicked. */
  onMarkerClick: () => void;

  selected?: IPropertyDetail | null;
};

/**
 * Get a new instance of a BBox from the specified 'bounds'.
 * @param bounds The latitude longitude boundary.
 */
const getBbox = (bounds: LatLngBounds): BBox => {
  return [
    bounds.getSouthWest().lng,
    bounds.getSouthWest().lat,
    bounds.getNorthEast().lng,
    bounds.getNorthEast().lat,
  ] as BBox;
};

interface ITilePoint {
  // x axis of the tile
  x: number;
  // y axis of the tile
  y: number;
  // zoom state of the tile
  z: number;
}

interface ITile {
  // Tile point {x, y, z}
  point: ITilePoint;
  // unique id of the file
  key: string;
  // bbox of the tile
  bbox: string;
  // tile data status
  processed?: boolean;
  // tile data, a list of properties in the tile
  datum?: PointFeature[];
  // tile bounds
  latlngBounds: LatLngBounds;
}

/**
 * Generate tiles for current bounds and zoom
 * @param bounds
 * @param zoom
 */
export const getTiles = (bounds: LatLngBounds, zoom: number): ITile[] => {
  const bbox = {
    bottom: bounds.getSouth(),
    left: bounds.getWest(),
    top: bounds.getNorth(),
    right: bounds.getEast(),
  };

  const tiles = tilesInBbox(bbox, zoom);

  // convert tile x axis to longitude
  const tileToLong = (x: number, z: number) => {
    return (x / Math.pow(2, z)) * 360 - 180;
  };

  // convert tile y axis to longitude
  const tileToLat = (y: number, z: number) => {
    const n = Math.PI - (2 * Math.PI * y) / Math.pow(2, z);

    return (180 / Math.PI) * Math.atan(0.5 * (Math.exp(n) - Math.exp(-n)));
  };

  return tiles.map(({ x, y, z }) => {
    const SW_long = tileToLong(x, z);

    const SW_lat = tileToLat(y + 1, z);

    const NE_long = tileToLong(x + 1, z);

    const NE_lat = tileToLat(y, z);

    return {
      key: `${x}:${y}:${z}`,
      bbox: SW_long + ',' + SW_lat + ',' + NE_long + ',' + NE_lat + ',EPSG:4326',
      point: { x, y, z },
      datum: [],
      latlngBounds: new LatLngBounds({ lat: SW_lat, lng: SW_long }, { lat: NE_lat, lng: NE_long }),
    };
  });
};

// default BC map bounds
export const defaultBounds = new LatLngBounds(
  [60.09114547, -119.49609429],
  [48.78370426, -139.35937554],
);

/**
 * Displays the search results onto a layer with clustering.
 * This component makes a request to the PIMS API properties search WFS endpoint.
 */
export const InventoryLayer: React.FC<InventoryLayerProps> = ({
  bounds,
  zoom,
  filter,
  onMarkerClick,
  selected,
  onRequestData,
}) => {
  const mapInstance = useMap();
  const [features, setFeatures] = useState<PointFeature[]>([]);
  const [loadingTiles, setLoadingTiles] = useState(false);
  const { loadProperties } = useApi();
  const { changed: filterChanged } = useFilterContext();

  const draftProperties: PointFeature[] = useAppSelector(
    state => state.properties?.draftProperties ?? [],
  );

  if (!mapInstance) {
    throw new Error('<InventoryLayer /> must be used under a <Map> leaflet component');
  }

  const bbox = useMemo(() => getBbox(bounds), [bounds]);
  useEffect(() => {
    const fit = async () => {
      if (filterChanged) {
        mapInstance.fitBounds(defaultBounds, { maxZoom: 5 });
      }
    };

    fit();
  }, [mapInstance, filter, filterChanged]);

  const params = useMemo((): any => {
    const tiles = getTiles(defaultBounds, 5);

    return tiles.map(tile => ({
      STREET_ADDRESS_1: filter?.STREET_ADDRESS_1,
      PID: filter?.PID,
      PIN: filter?.PIN,
      BBOX: tile.bbox,
    }));
  }, [filter]);

  const loadTile = async (mapFilter: IGeoSearchParams) => {
    return loadProperties(mapFilter);
  };

  /**
   * TODO: convert polygon features into lat/lng coordinates, remove this when new Point Geoserver layer available.
   * @param feature the feature to obtain lat/lng coordinates for.
   * @returns [lat, lng]
   */
  const getLatLng = (feature: any) => {
    if (feature.geometry.type === 'Polygon' || feature.geometry.type === 'MultiPolygon') {
      const latLng = geoJSON(feature.geometry)
        .getBounds()
        .getCenter();
      return [latLng.lng, latLng.lat];
    }
    return feature.geometry.coordinates;
  };

  const search = async (filters: IGeoSearchParams[]) => {
    //TODO: currently this loads all matching properties, this should be rewritten to use the bbox and make one request per tile.
    try {
      onRequestData(true);
      const tileData = await loadTile(filters[0]);
      const validFeatures = tileData.features.filter(feature => !!feature?.geometry);
      const data = validFeatures.map((feature: Feature) => {
        //TODO: this converts all polygons to points, this should be changed to a View that returns the POINT instead of the POLYGON (psp-1859)
        return {
          ...feature,
          geometry: { type: 'Point', coordinates: getLatLng(feature) },
          properties: {
            ...feature.properties,
            propertyTypeId: PropertyTypes.Land,
          },
        } as Feature;
      });

      const results = uniqBy(data, (point: Feature) => `${point?.properties?.PROPERTY_ID}`);

      setFeatures(results as any);
      setLoadingTiles(false);
      if (results.length === 0) {
        toast.info('No search results found');
      } else {
        toast.info(`${results.length} properties found`);
      }
    } catch (error) {
      toast.error((error as Error).message, { autoClose: 7000 });
    } finally {
      onRequestData(false);
    }
  };

  useMapRefreshEvent(() => search(params));
  useDeepCompareEffect(() => {
    setLoadingTiles(true);
    search(params);
  }, [params]);

  return (
    <PointClusterer
      points={features}
      draftPoints={draftProperties}
      zoom={zoom}
      bounds={bbox}
      onMarkerClick={onMarkerClick}
      zoomToBoundsOnClick={true}
      spiderfyOnMaxZoom={true}
      selected={selected}
      tilesLoaded={!loadingTiles}
    />
  );
};
