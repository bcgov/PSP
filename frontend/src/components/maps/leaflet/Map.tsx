import './Map.scss';

import axios from 'axios';
import { useLayerQuery } from 'components/maps/leaflet/LayerPopup';
import { IGeoSearchParams } from 'constants/API';
import { MAP_MAX_ZOOM } from 'constants/strings';
import { PropertyFilter } from 'features/properties/filter';
import { IPropertyFilter } from 'features/properties/filter/IPropertyFilter';
import { Feature } from 'geojson';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import { geoJSON, LatLng, LatLngBounds, LeafletMouseEvent, Map as LeafletMap } from 'leaflet';
import isEmpty from 'lodash/isEmpty';
import isEqual from 'lodash/isEqual';
import isEqualWith from 'lodash/isEqualWith';
import React, { useEffect, useRef, useState } from 'react';
import Container from 'react-bootstrap/Container';
import { MapContainer as ReactLeafletMap, Popup, TileLayer } from 'react-leaflet';
import { useDispatch } from 'react-redux';
import { useResizeDetector } from 'react-resize-detector';
import { useMediaQuery } from 'react-responsive';
import { useAppSelector } from 'store/hooks';
import { DEFAULT_MAP_ZOOM, setMapViewZoom } from 'store/slices/mapViewZoom/mapViewZoomSlice';
import { saveParcelLayerData } from 'store/slices/parcelLayerData/parcelLayerDataSlice';
import { IPropertyDetail, storeProperty } from 'store/slices/properties';

import { Claims } from '../../../constants';
import BasemapToggle, { BaseLayer, BasemapToggleEvent } from '../BasemapToggle';
import useActiveFeatureLayer from '../hooks/useActiveFeatureLayer';
import { useFilterContext } from '../providers/FIlterProvider';
import { PropertyPopUpContextProvider } from '../providers/PropertyPopUpProvider';
import InfoSlideOut from './InfoSlideOut/InfoSlideOut';
import { InventoryLayer } from './InventoryLayer';
import {
  MUNICIPALITY_LAYER_URL,
  municipalityLayerPopupConfig,
  parcelLayerPopupConfig,
  PARCELS_LAYER_URL,
} from './LayerPopup/constants';
import {
  LayerPopupContent,
  LayerPopupTitle,
  PopupContentConfig,
} from './LayerPopup/LayerPopupContent';
import LayersControl from './LayersControl';
import { LegendControl } from './Legend/LegendControl';
import LoadingBackdrop from './LoadingBackdrop/LoadingBackdrop';
import { MapEvents } from './MapEvents/MapEvents';
import * as Styled from './styles';
import { ZoomOutButton } from './ZoomOut/ZoomOutButton';

export type MapViewportChangeEvent = {
  bounds: LatLngBounds | null;
  filter?: IGeoSearchParams;
};

export type MapProps = {
  lat: number;
  lng: number;
  zoom?: number;
  selectedProperty?: IPropertyDetail | null;
  onViewportChanged?: (e: MapViewportChangeEvent) => void;
  onMapClick?: (e: LeafletMouseEvent) => void;
  disableMapFilterBar?: boolean;
  interactive?: boolean;
  showParcelBoundaries?: boolean;
  whenCreated?: (map: LeafletMap) => void;
  whenReady?: () => void;
};

export type LayerPopupInformation = PopupContentConfig & {
  latlng: LatLng;
  title: string;
  center?: LatLng;
  bounds?: LatLngBounds;
  feature: Feature;
};

const defaultFilterValues: IPropertyFilter = {
  searchBy: 'pid',
  pid: '',
  pin: '',
  address: '',
  location: '',
};

const whitelistedFilterKeys = ['PID', 'PIN', 'ADDRESS', 'LOCATION'];

/**
 * Converts the map filter to a geo search filter.
 * @param filter The map filter.
 */
const getQueryParams = (filter: IPropertyFilter): IGeoSearchParams => {
  return {
    PID: filter.pid ? +filter.pid?.replace(/-/g, '') : undefined,
    PIN: filter.pin ? +filter.pin?.replace(/-/g, '') : undefined,
    STREET_ADDRESS_1: filter.address,
  };
};

const defaultBounds = new LatLngBounds([60.09114547, -119.49609429], [48.78370426, -139.35937554]);

/**
 * Creates a Leaflet map and by default includes a number of preconfigured layers.
 * @param param0
 */
const Map: React.FC<MapProps> = ({
  lat,
  lng,
  zoom: zoomProp,
  selectedProperty,
  onMapClick,
  disableMapFilterBar,
  whenReady,
  whenCreated,
}) => {
  const keycloak = useKeycloakWrapper();
  const dispatch = useDispatch();
  const [geoFilter, setGeoFilter] = useState<IGeoSearchParams>({
    ...defaultFilterValues,
    includeAllProperties: keycloak.hasClaim(Claims.ADMIN_PROPERTIES),
  } as any);
  const [baseLayers, setBaseLayers] = useState<BaseLayer[]>([]);
  const [triggerFilterChanged, setTriggerFilterChanged] = useState(true);
  const [activeBasemap, setActiveBasemap] = useState<BaseLayer | null>(null);
  const smallScreen = useMediaQuery({ maxWidth: 1800 });
  const municipalitiesService = useLayerQuery(MUNICIPALITY_LAYER_URL);
  const parcelsService = useLayerQuery(PARCELS_LAYER_URL);
  const [bounds, setBounds] = useState<LatLngBounds>(defaultBounds);
  const { setChanged } = useFilterContext();
  const [layerPopup, setLayerPopup] = useState<LayerPopupInformation>();

  // a reference to the internal Leaflet map instance (this is NOT a react-leaflet class but the underlying leaflet map)
  const mapRef = useRef<LeafletMap | null>(null);

  if (mapRef.current && !selectedProperty?.propertyDetail) {
    const center = mapRef.current.getCenter();
    lat = center.lat;
    lng = center.lng;
  }

  const parcelLayerFeature = useAppSelector(state => state.parcelLayerData?.parcelLayerFeature);
  useActiveFeatureLayer({
    selectedProperty,
    layerPopup,
    mapRef,
    parcelLayerFeature,
    setLayerPopup,
  });
  const [showLoadingBackdrop, setShowLoadingBackdrop] = useState(true);

  const lastZoom = useAppSelector(state => state.mapViewZoom) ?? zoomProp;
  const [zoom, setZoom] = useState(lastZoom);
  useEffect(() => {
    if (lastZoom === DEFAULT_MAP_ZOOM) {
      dispatch(setMapViewZoom(smallScreen ? 4.9 : 5.5));
    } else if (lastZoom !== zoom && zoom !== DEFAULT_MAP_ZOOM) {
      dispatch(setMapViewZoom(zoom));
    }
  }, [dispatch, lastZoom, smallScreen, zoom]);

  const { width, ref: resizeRef } = useResizeDetector();
  useEffect(() => {
    mapRef.current?.invalidateSize();
  }, [mapRef, width]);

  const handleMapFilterChange = async (filter: IPropertyFilter) => {
    const compareValues = (objValue: any, othValue: any) => {
      return whitelistedFilterKeys.reduce((acc, key) => {
        return (isEqual(objValue[key], othValue[key]) || (!objValue[key] && !othValue[key])) && acc;
      }, true);
    };
    // Search button will always trigger filter changed (triggerFilterChanged is set to true when search button is clicked)
    if (!isEqualWith(geoFilter, getQueryParams(filter), compareValues) || triggerFilterChanged) {
      dispatch(storeProperty(null));
      setGeoFilter(getQueryParams(filter));
      setChanged(true);
      setTriggerFilterChanged(false);
    }
  };

  const handleBasemapToggle = (e: BasemapToggleEvent) => {
    const { previous, current } = e;
    setBaseLayers([current, previous]);
    setActiveBasemap(current);
  };

  useEffect(() => {
    // fetch GIS base layers configuration from /public folder
    axios.get('/basemaps.json')?.then(result => {
      setBaseLayers(result.data?.basemaps);
      setActiveBasemap(result.data?.basemaps?.[0]);
    });
  }, []);

  const fitMapBounds = () => {
    if (mapRef.current) {
      mapRef.current.fitBounds([
        [60.09114547, -119.49609429],
        [48.78370426, -139.35937554],
      ]);
    }
  };

  const handleMapReady = () => {
    fitMapBounds();
    if (typeof whenReady === 'function') {
      whenReady();
    }
  };

  const handleMapCreated = (mapInstance: L.Map) => {
    mapRef.current = mapInstance;
    if (typeof whenCreated === 'function') {
      whenCreated(mapInstance);
    }
  };

  const showLocationDetails = async (event: LeafletMouseEvent) => {
    !!onMapClick && onMapClick(event);
    const municipality = await municipalitiesService.findOneWhereContains(event.latlng);
    const parcel = await parcelsService.findOneWhereContains(event.latlng);

    let properties = {};
    let center: LatLng | undefined;
    let mapBounds: LatLngBounds | undefined;
    let displayConfig = {};
    let title = 'Municipality Information';
    let feature = {};
    if (municipality?.features?.length === 1) {
      properties = municipality.features[0].properties!;
      displayConfig = municipalityLayerPopupConfig;
      feature = municipality.features[0];
      mapBounds = municipality.features[0]?.geometry
        ? geoJSON(municipality.features[0].geometry).getBounds()
        : undefined;
    }

    if (parcel?.features?.length === 1) {
      title = 'Parcel Information';
      properties = parcel.features[0].properties!;
      displayConfig = parcelLayerPopupConfig;
      mapBounds = parcel.features[0]?.geometry
        ? geoJSON(parcel.features[0].geometry).getBounds()
        : undefined;
      center = mapBounds?.getCenter();
      feature = parcel.features[0];
    }

    if (!isEmpty(properties)) {
      setLayerPopup({
        title,
        data: properties as any,
        config: displayConfig as any,
        latlng: event.latlng,
        center,
        bounds,
        feature,
      } as any);
    }
  };

  const handleBounds = (e: any) => {
    const boundsData: LatLngBounds = e.target.getBounds();
    if (!isEqual(boundsData.getNorthEast(), boundsData.getSouthWest())) {
      setBounds(boundsData);
    }
  };

  const property = useAppSelector(state => state.properties?.propertyDetail);
  const [infoOpen, setInfoOpen] = React.useState(!!property);
  const [layersOpen, setLayersOpen] = React.useState(false);

  return (
    <Styled.MapGrid ref={resizeRef} className="px-0 map">
      <LoadingBackdrop show={showLoadingBackdrop} />
      {!disableMapFilterBar ? (
        <Container fluid className="px-0 map-filter-container">
          <PropertyFilter
            defaultFilter={{
              ...defaultFilterValues,
            }}
            onChange={handleMapFilterChange}
            setTriggerFilterChanged={setTriggerFilterChanged}
          />
        </Container>
      ) : null}
      <Styled.MapContainer>
        {baseLayers?.length > 0 && (
          <BasemapToggle baseLayers={baseLayers} onToggle={handleBasemapToggle} />
        )}
        <PropertyPopUpContextProvider>
          <ReactLeafletMap
            center={[lat, lng]}
            zoom={lastZoom}
            maxZoom={MAP_MAX_ZOOM}
            closePopupOnClick={true}
            whenCreated={handleMapCreated}
            whenReady={handleMapReady}
          >
            <MapEvents
              click={showLocationDetails}
              zoomend={e => setZoom(e.sourceTarget.getZoom())}
              moveend={handleBounds}
            />
            {activeBasemap && (
              <TileLayer
                attribution={activeBasemap.attribution}
                url={activeBasemap.url}
                zIndex={0}
              />
            )}
            {!!layerPopup && (
              <Popup
                position={layerPopup.latlng}
                offset={[0, -25]}
                onClose={() => {
                  setLayerPopup(undefined);
                  dispatch(storeProperty(null));
                }}
                closeButton={true}
                autoPan={false}
              >
                <LayerPopupTitle>{layerPopup.title}</LayerPopupTitle>
                <LayerPopupContent
                  data={layerPopup.data as any}
                  config={layerPopup.config as any}
                  center={layerPopup.center}
                  onAddToParcel={(e: MouseEvent, data: { [key: string]: any }) => {
                    dispatch(
                      saveParcelLayerData({
                        e: { timeStamp: document?.timeline?.currentTime ?? 0 } as any,
                        data: {
                          ...data,
                          CENTER: { lat: data?.CENTER.lat, lng: data?.CENTER.lng },
                        },
                      }),
                    );
                  }}
                  bounds={layerPopup.bounds}
                />
              </Popup>
            )}
            <LegendControl />
            <ZoomOutButton bounds={defaultBounds} />
            <LayersControl
              open={layersOpen}
              setOpen={() => {
                setLayersOpen(!layersOpen);
                setInfoOpen(false);
              }}
            />
            <InfoSlideOut
              open={infoOpen}
              setOpen={(state: boolean) => {
                setInfoOpen(state);
                setLayersOpen(false);
              }}
              onHeaderActionClick={() => {
                setInfoOpen(false);
              }}
            />
            <InventoryLayer
              zoom={zoom}
              bounds={bounds}
              onMarkerClick={() => {
                if (!infoOpen) {
                  setLayersOpen(false);
                  setInfoOpen(true);
                }
              }}
              selected={selectedProperty}
              filter={geoFilter}
              onRequestData={setShowLoadingBackdrop}
            ></InventoryLayer>
          </ReactLeafletMap>
        </PropertyPopUpContextProvider>
      </Styled.MapContainer>
    </Styled.MapGrid>
  );
};

export default Map;
