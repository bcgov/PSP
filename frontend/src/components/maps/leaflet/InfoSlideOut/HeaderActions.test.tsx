import 'leaflet';
import 'leaflet/dist/leaflet.css';

import { useKeycloak } from '@react-keycloak/web';
import { render } from '@testing-library/react';
import { PropertyTypes } from 'constants/propertyTypes';
import { createMemoryHistory } from 'history';
import { IProperty } from 'interfaces';
import { noop } from 'lodash';
import { mockParcel } from 'mocks/filterDataMock';
import * as React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import HeaderActions from './HeaderActions';

jest.mock('@react-keycloak/web');
(useKeycloak as jest.Mock).mockReturnValue({
  keycloak: {
    userInfo: {
      organizations: [1],
      roles: [],
    },
    subject: 'test',
  },
});

const history = createMemoryHistory();
const mockStore = configureMockStore([thunk]);
const store = mockStore({});

const HeaderComponent = (
  propertyInfo: IProperty | null,
  propertyTypeId: PropertyTypes | null,
  canViewDetails: boolean,
  canEditDetails: boolean,
) => {
  return (
    <Provider store={store}>
      <Router history={history}>
        <HeaderActions
          propertyInfo={propertyInfo}
          propertyTypeId={propertyTypeId}
          canViewDetails={canViewDetails}
          canEditDetails={canEditDetails}
          zoomToView={noop}
        />
      </Router>
    </Provider>
  );
};

describe('HeaderActions View', () => {
  it('HeaderActions renders correctly', () => {
    const { container } = render(HeaderComponent(mockParcel, PropertyTypes.Land, true, true));
    expect(container.firstChild).toMatchSnapshot();
  });

  xit('Shows all three actions when user has all permissions', () => {
    const { getByText } = render(HeaderComponent(mockParcel, PropertyTypes.Land, true, true));
    expect(getByText('Zoom map')).toBeVisible();
  });
});
