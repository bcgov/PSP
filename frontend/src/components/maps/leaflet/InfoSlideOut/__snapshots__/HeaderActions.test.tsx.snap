// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderActions View HeaderActions renders correctly 1`] = `
.c0 {
  background-color: filterBackgroundColor;
  height: 3.5rem;
  width: 32.2rem;
  margin: 0rem 0rem 0.5rem -1rem;
  font-size: 1.4rem;
  padding: 1rem;
}

.c0 a {
  padding: 0rem 1rem;
  color: slideOutBlue;
}

<div
  class="c0 row"
>
  Actions:
  <a
    href="/"
  >
    Zoom map
  </a>
</div>
`;
